frame/chaos/src/lib.rs

```diff
@@ -0,0 +1,501 @@
+			T::ensure_protocol_nft_owner::<ChaosStakingNFTOf<T>>(&owner, &instance_id)?;
+			<Self as ChaosProtocol>::claim(&to, &instance_id)?;
+			Ok(().into())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// The chaos protocol account. Derived from the chaos pallet id.
+		pub(crate) fn account_id() -> AccountIdOf<T> {
+			T::PalletId::get().into_account()
+		}
+
+		/// Current reward indexes.
+		pub(crate) fn current_reward_indexes() -> ChaosStateOf<T> {
+			RewardAssets::<T>::get()
+				.into_iter()
+				.map(|x| (x, RewardIndexes::<T>::get(&x).unwrap_or(RewardIndex::zero())))
+				.collect::<BTreeMap<_, _>>()
+				.try_into()
+				.expect("map does not alter the length; qed;")
+		}
+	}
+
+	impl<T: Config> ChaosProtocol for Pallet<T> {
+		type AccountId = AccountIdOf<T>;
+		type AssetId = AssetIdOf<T>;
+		type Balance = BalanceOf<T>;
+		type InstanceId = T::InstanceId;
+
+		fn stake(
+			from: &Self::AccountId,
+			amount: Self::Balance,
+			duration: Timestamp,
+			keep_alive: bool,
+		) -> Result<Self::InstanceId, DispatchError> {
+			// Make sure the duration is a valid preset.
+			ensure!(Self::staking_presets().contains(&duration), Error::<T>::InvalidStakingPreset);
+
+			// Acquire Chaos from user.
+			let chaos_account = Self::account_id();
+			T::ChaosAsset::transfer(from, &chaos_account, amount, keep_alive)?;
+
+			// Actually create the NFT representing the user position.
+			let reward_indexes = Self::current_reward_indexes();
+			let now = T::Time::now();
+			let nft = ChaosStakingNFT {
+				stake: amount,
+				reward_indexes,
+				lock_date: now.as_secs(),
+				lock_duration: duration,
+			};
+			let instance_id = T::mint_protocol_nft(from, &nft)?;
+
+			// Increment total shares.
+			TotalShares::<T>::try_mutate(|x| {
+				x.checked_add(amount.saturated_into::<u128>().into())
+					.ok_or(ArithmeticError::Overflow)
+			})?;
+
+			// Trigger event
+			Self::deposit_event(Event::<T>::Staked {
+				who: from.clone(),
+				stake: amount,
+				nft: instance_id,
+			});
+
+			Ok(instance_id)
+		}
+
+		fn unstake(to: &Self::AccountId, instance_id: &Self::InstanceId) -> DispatchResult {
+			// Make sure we execute a final claim before unstaking.
+			<Self as ChaosProtocol>::claim(to, instance_id)?;
+
+			let nft = T::get_protocol_nft::<ChaosStakingNFTOf<T>>(instance_id)?;
+
+			let now = T::Time::now();
+
+			// Possibly penalize the staked asset, Chaos in this case.
+			let (penalized_stake, penalty_amount) =
+				nft.penalize_early_unstake_amount(now.as_secs(), EarlyUnstakePenalty::<T>::get())?;
+
+			// Transfer back the (possibly penalized) staked amount.
+			let chaos_account = Self::account_id();
+			// NOTE(hussein-aitlahcen): no need to keep protocol account alive.
+			T::ChaosAsset::transfer(&chaos_account, &to, penalized_stake, false)?;
+
+			// Decrement total shares.
+			// TODO: decrement by penalized only????? the protocol would own this fraction
+			// indefinitely
+			TotalShares::<T>::try_mutate(|x| {
```

Comment:
decremente by original value, you brought amount and you took amount. simmetrical. you burn - so to make others rich faster.  if you burn not whole amount you make inflation to happen faster. it feels that burning is good.